### DISCUSSION TITLE
Case sensitivity global option for better cross-platform support

### DISCUSSION THREAD

**Discussion by Travisyard** [beetbox/beets#5104] [Ideas]
Hello beets community,

My music library's files are synced to multiple systems with different operating systems and filesystem formats. My beets library is managed on Linux which is usually case-sensitive. Some platforms (namely Android and Windows) are usually case-insensitive, and this has caused me some headaches.

For example, one artist I have in my library has released albums as both "Diaf" and "DIAF". On most Linux systems, these two directory names can exist side-by-side. However, on Windows and Android they are generally considered the same directory. Another example of this occurred when I tried to replace a corrupted file in the same album. The original album folder was "Sound **o**f Silver", but the replacement song imported into the folder "Sound **O**f Silver". This resulted in the same failure mode.

One workaround I have found is to use the `%upper{}`, `%lower{}`, or `%title{}` functions in the path formats, but this workaround is only bulletproof if every `$template` substitution is nested inside one of these functions. Also, I'd like to retain filenames that match the preferred stylization of the artists I listen to ("ABSRDST" and "LOOK MUM NO COMPUTER" should be all uppercase, "ev.exi" and "clipping." should be all lowercase).

I suggest adding a global configuration option like `case_sensitivity: yes|no`. The default for systems that are usually case-insensitive should be `no`, and the default for systems that are usually case-sensitive should be `yes`. Users who manage their beets library on a case-sensitive system and then copy/move/synchronize to a case-insensitive system would be able to change this option to prevent unexpected copy/move/synchronize errors.

Reference: [Case Sensitivity on Wikipedia](https://en.wikipedia.org/wiki/Case_sensitivity)

**Comment by Serene-Arc** [comment]
Sorry, just need some clarification here. What is your preferred behaviour? What do you think that beets should do instead of what it is doing?

As far as beets as concerned, the artist is different if the case is different, unless, as you said, you put it in a template function that changes it. I'm a little confused as to what you think beets should be doing. You're saying that you want the preferred style of the artist but seems mutually incompatible with not having case sensitive names.

**Reply by Travisyard** [reply]
Currently, the directory structure below is a possible output from beets where MusicBrainz and/or the artist in question are inconsistent with their capitalization:

~/music/DIAF/Gloria
~/music/Diaf/Weida
~/music/LCD Soundsystem/Sound Of Silver
~/music/LCD Soundsystem/Sound of Silver

This directory structure is valid on Linux. The problem is that the above directory structure is not valid on Windows or Android which introduces interoperability barriers for users of multiple OSes. For example, attempting to synchronize the above with Syncthing can produce unexpected errors like `remote “music/DIAF” uses different upper or lowercase characters than local “music/Diaf”; change the casing on either side to match the other`.

Now, the desired behavior. Let's assume that the directory structure already looks like this with only Weida and Sound Of Silver imported:

~/music/Diaf/Weida
~/music/LCD Soundsystem/Sound Of Silver

If `case_sensitivity` is set to `no`, when I import Gloria or replace a song on Sound Of Silver which have different case styles beets should NOT create additional, differently-cased directories. Instead, the directory structure remains with just these two folders. The associated metadata in the beets database should retain the different case styles from the import. This would be the final, desired directory structure:

~/music/Diaf/Gloria
~/music/Diaf/Weida
~/music/LCD Soundsystem/Sound Of Silver

Of course, if I had imported Gloria first instead of Weida, or if the metadata source I used for the initial import of LCD Soundsystem's album had the lowercase "of", the directory structure would look end up looking like this...

~/music/DIAF/Gloria
~/music/DIAF/Weida
~/music/LCD Soundsystem/Sound of Silver

...which is also entirely valid and acceptable as it does not pass case-sensitivity errors to downstream processes, copies, or synchronizations. Please correct me if I'm wrong, but I believe this is how Windows handles differently-cased directories as it is case-insensitive but also case-_preserving_.

**Reply by RollingStar** [reply]
IDK the most performant way to do this, but you're suggesting checking if a case insensitive match exists for all parent folders, and only making new folders/subfolders where they don't. Pathlib appears to have an interface for this: https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.match

There must be a cross platform project that has solved this problem, like FTP or rsync.

**Reply by Serene-Arc** [reply]
I'll have to take a look at the specific code in the modules as to whether this would be a difficult change to make. @Travisyard if you'd like to make a bug report, we can get the ball rolling on that.

I want to be clear though. Beets considers the albums 'gloria' and 'Gloria' to be different because they do not have the same name. By collapsing that, it is possible (however unlikely) that losing that information will lead to errors. Consider a band that releases two albums, one of which is a new edition that is signified by a different case e.g. 'first album' vs 'FIRST album'. Those directories would be merged and treated as the same on the file system.

**Reply by Travisyard** [reply]
@Serene-Arc I didn't file a bug report because it isn't really a bug. The unexpected behavior I have encountered occurred when I passed the files to a downstream process. With that in mind, do you still want me to file a bug report?

Also re: case disambiguation, that is a great point! With that said, perhaps a small modification to my original proposal will bring it inline with that limitation. I am not proposing any changes at all to how Beets catalogues things in its internal database. It can continue to be internally case-sensitive. The changes I am proposing _only affect directory structure_. Internally, "Gloria" and "gloria" can be different albums, but would then be disambiguated on the disk as something like "Artist/Gloria" and "Artist/gloria-1". Do you think this is more feasible than trying to automatically merge them as I had previously proposed?

**Reply by Serene-Arc** [reply]
Unexpected behaviour can still be a bug if we need to add it to the documentation explaining what it is. Unexpected behaviour, whether from a compiler or from the perspective of the user, is something that programs should avoid. Plus we also have feature requests. Having it under a bug report of feature request makes it far easier to track and assign work compared to a discussion.

That would be more conceptually feasible, though it might run counter to what the user expects it to do. I'm not overly familiar with that part of the code though.

I should note that this is concerning paths and I intend to overhaul the path code of beets soon :tm: . I keep saying that, but maybe this will be a reason to push forwards with that. As @RollingStar mentioned, pathlib can deal with this issue and once we switch the internal code to pathlib, it might be easier to implement this feature.

**Reply by Travisyard** [reply]
@Serene-Arc I have submitted [this issue](https://github.com/beetbox/beets/issues/5113) based on our discussion here. Thank you for your insight and help with this!
